### ISSUE TITLE
Memory leak when using DocumentClient?

### ISSUE THREAD

**Issue by AranVinkItility** [Azure/azure-cosmosdb-node#198]
I'm using the DocumentClient in production with Azure Functions, it all works fine, but I'm seeing an increase in memory consumption over time.

To test what was causing this I replaced all of my external calls with stubs, then reran the application for a longer time. No memory leak occurred.

Only after re-enabling the DocumentDB client my memory consumption started to rise.

Using heapdump, I captured a memory dump before and after doing 200 requests to the DocumentDB using the client:
![image](https://user-images.githubusercontent.com/16576319/26923361-28bc842e-4c42-11e7-9649-34bb85a98b2b.png)

Comparing the two I noticed that TLSSockets do not seem to garbage collect properly and keep using memory.

Did anyone else notice this behavior, or have any pointers on how to further investigate this?

**Comment by moderakh** [comment]
@AranVinkItility 
Could you please share the code to reproduce this (which queries are you using, etc)? I will have someone to look at this.

**Comment by AranVinkItility** [comment]
@moderakh Would really appreciate it, here is my code:
```javascript
'use strict';
const heapdump = require('heapdump');
const Promise = require('bluebird');
const DocumentClient = require('documentdb').DocumentClient;

const host = 'https://yourhost.documents.azure.com';
const key = 'yourkey';
const dbName = 'dbnamewithcollections';

const client = new DocumentClient(host, { masterKey: key });

function listCollections(databaseLink) {
    return Promise.fromCallback(function (callback) {
        client.readCollections(databaseLink).toArray(callback);
    });
}

const numberOfRequests = 100;
let counterCompleted = 0;
let promises = [];

for (var x = 0; x < numberOfRequests; x++) {
    promises.push(
        listCollections('dbs/' + dbName)
    );
}

console.log('Writing snapshot (before calling docDB)');
heapdump.writeSnapshot();
Promise.all(promises)
    .then(function (result) {
        console.log(result.length);
    })
    .then(function () {
        console.log('Writing snapshot (after calling docDB)');
        heapdump.writeSnapshot();
    });
```

**Comment by moderakh** [comment]
@mehdijazayeri3000  can you please take a look at this?

**Comment by mehdijazayeri3000** [comment]
That's probably because toArray() implementation. Before we release the new version, please 
go to queryIterator.js. file and find_toArrayImplementation function. Change this line
"**that.toArrayTempResources = that.toArrayTempResources.concat(resource);**" 
with "**that.toArrayTempResources.push(resource);**"
Please let us know if that solved your problem.

**Comment by itilitymaarten** [comment]
@mehdijazayeri3000 I'm a colleague of @AranVinkItility, I'll be taking over this issue in his absence.

I have tried your suggestion but I don't see an improvement, the heap size at the end of the function run stays the same (give or take a few KB, of course) as it was before the modification.

**Comment by mehdijazayeri3000** [comment]
@AranVinkItility Testing memory leak in node is a bit tricky. V8 GC is not a reference-counting GC. Instead, it does a full mark-and-sweep GC. It won't start cleaning up unless we get over a certain amount of memory. Please read this <a href = 'https://strongloop.com/strongblog/node-js-performance-garbage-collection/'>post </a> for more information.

**Comment by itilitymaarten** [comment]
@mehdijazayeri3000 I agree that we can't easily conclude there is a memory leak, however I still have my suspicions. I've looked into forcing garbage collection, and came up with the following code:

```
'use strict';
const heapdump = require('heapdump');

function generateData() {
    const host = '<host-url>';
    const key = '<authorization-key>';
    const dbName = '<database-name>';

    let DocumentClient = require('documentdb').DocumentClient;
    let client = new DocumentClient(host, { masterKey: key });

    global.gc();
    console.log('Writing snapshot (before starting loop)');
    heapdump.writeSnapshot();

    for (var x = 0; x < 100; x++) {
        client.readCollections('dbs/' + dbName).toArray(() => {});
    }

    client = null;
    DocumentClient = null;
}

generateData();

setTimeout(() => {
    global.gc();
    console.log('Writing snapshot (after timeout)');
    heapdump.writeSnapshot();
}, 5000);
```
You have to run this with `node --expose-gc` to enable the `global.gc()` function.

I've tried every trick I know to ensure the DocumentClient instance and all other data goes out of scope, calling the garbage collector and heapdump on a timeout to ensure it happens after the http calls without tieing directly to them. I do see some result from garbage collection, but it's very minimal; a huge amount of strings and other data is still floating around.

**Comment by southpolesteve** [comment]
We have recently announced deprecation version 1.x of the Azure Cosmos JavaScript SDK. We will end support for the `documentdb` package and this repo on August 30, 2020. Please update to our new package [@azure/cosmos](https://www.npmjs.com/package/@azure/cosmos) as soon as possible. If you encounter any issues, you can raise them in the [Azure central JS SDK repo](https://github.com/Azure/azure-sdk-for-js). If something is preventing you from upgrading to the latest version of the SDK, you can always email me directly: stfaul@microsoft.com
